Correct MixUntilAnonymitySet when loading Wasabi or when saving the settings

yahiheb opened the pull request:
Fixes https://github.com/zkSNACKs/WalletWasabi/issues/2711

https://github.com/zkSNACKs/WalletWasabi/issues/3494#issuecomment-623345996

It addresses https://github.com/zkSNACKs/WalletWasabi/issues/3494#issuecomment-621309977 so it may help fix that issue.

yahiheb commented:
> How about doing the corrections when the user messes around with the privacy level?

Done. https://github.com/zkSNACKs/WalletWasabi/pull/3616/commits/a6c01d625ed9b39d5ebc7fe3750758cb0299e4c2

> Also, you could add the correction that the privacy levels should be in an increasing order into CorrectMixUntilAnonymitySetValue.

I think we should do that in a followup PR.

molnard commented:
How about using the "CorrectMixUntilAnonymitySetValue" there instead of code duplication?

yahiheb commented:
> How about using the "CorrectMixUntilAnonymitySetValue" there instead of code duplication?

I think it doesn't work:
lets say the privacy levels are as following 2, 21, 50 and `MixUntilAnonymitySet` is 50
if the user changes them to 60, 70, 80 then `CorrectMixUntilAnonymitySetValue` will set `MixUntilAnonymitySet` to `PrivacyLevelSome` https://github.com/zkSNACKs/WalletWasabi/pull/3616/files#diff-8486af14b60ef042d533c6ad5c64b64cR338

molnard commented:
I was thinking about this. No matter what we do it will be just a "patch". I vote for changing MixUntil to an enum and that will simplify things here. 

This change will bubble up until the Target anonset where we have to find out what to do with currently enqueued coins - probably nothing but have to investigate. 

The current solution is to not touch it until restart but that will be broken with this change or already broken.

molnard commented:
Dev meeting: 

This PR was good because it highlighted the problem around not having an enum instead of MixUntilAnonSet. This design flaw comes, again and again, causing bugs. So even if there will be some "compatibility" code we decided to refactor this part instead of patching. This decision makes this PR obsolete. I am not closing this until we have another PR with the refactorings. @yahiheb are u available for that in a pair programming session?

yahiheb commented:
> @yahiheb are u available for that in a pair programming session?

Yes

lontivero commented on WalletWasabi.Gui/Config.cs:
```diff
@@ -329,12 +329,8 @@ public EndPoint GetP2PEndpoint()
 		}
 
-		public static Config LoadOrCreateDefaultFile(string path)
+		public static void CorrectMixUntilAnonymitySetValue(Config config)
```
I don't understand why this is `static`.

yahiheb commented:
@lontivero Sorry about the confusion.
This PR should be closed by #3645

> For example:
> 
> 1. Make sure all the values for all the privacy settings are okay and make sense. Try with Some = 2, Fine = 10 and Strong = 50. Make sure MixUntilAnonymitySet = 2.
> 2. Open Wasabi and open a wallet file.
> 3. Go to Tools > Settings and change Strong to 40 (check the Config.json file contains the change)
> 4. Go to Coinjoin tab and click the Target button.
> 5. The values in the config File are reverted and Strong is 50 again.

Me and @molnard discussed this is issue and it is a bit complicated (#3645 doesn't fix it)

> The behaviors become even more weird if the privacy levels make no sense like Some = 5 Fine = 1 and Strong = 50 or other values.

That is true. #2712 fixes that.

lontivero commented:
Sorry, I don't understand how these 3 PRs work together. I see the #3645 also touches the same methods. If this PR has to be closed please do it.

yahiheb commented:
- This PR was supposed to correct MixUntilAnonymitySet value if it is not equal to any of the privacy levels.
For example if the levels are 2, 21, 50 and MixUntilAnonymitySet is 60 it should be corrected to 50.


- With #3645 we store MixUntilAnonymitySet as a string that indicates which privacy level it is (this way we make sure it is always equal to one of the privacy levels)
```
"MixUntilAnonymitySet": "PrivacyLevelStrong",
"PrivacyLevelSome": 2,
"PrivacyLevelFine": 21,
"PrivacyLevelStrong": 50,
```


- #2712 makes sure that PrivacyLevelSome < PrivacyLevelFine < PrivacyLevelStrong
it addresses your comment:
`The behaviors become even more weird if the privacy levels make no sense like Some = 5 Fine = 1 and Strong = 50 or other values.
`

---
If those PRs get merged we will have only one problem left with the MixUntilAnonymitySet which is the one you described in your example above.
